### PR TITLE
fix: type check for constants

### LINKED ISSUE #2592: typechecker does not check constants
### Version Information

* vyper Version (output of `vyper --version`): v0.3.1

### What's your issue about?

During constant folding, the annotated types for constants do not get propagated. Instead they are translated into literals and the literal types are inferred at the use site. For example the following example compiles even though it shouldn't really:
```python
MY_CONSTANT: constant(uint256[3]) = [1,2,3]

@external
def foo() -> uint8[3]:
    return MY_CONSTANT
```

Compare the error generated for the following

```python
@external
def foo() -> uint8[3]:
    xs: uint256[3] = [1,2,3]
    return xs
```
```
Error compiling: tmp/foo.vy
vyper.exceptions.TypeMismatch: Given reference has type uint256[3], expected uint8[3]
  contract "tmp/foo.vy", function "foo", line 7:11 
       6     xs: uint256[3] = [1,2,3]
  ---> 7     return xs
  ------------------^
       8
```

### How can it be fixed?

Propagate type annotations during constant folding

### PR DESCRIPTION
**Pull request by tserg**
### What I did

Fix #2592.

Based on the issue, the following code compiles when it should not:
```
MY_CONSTANT: constant(uint256[3]) = [1,2,3]

@external
def foo() -> uint8[3]:
    return MY_CONSTANT
```

While looking into this, I discovered two other similar code snippets that also compile when they should not:
```
MY_CONSTANT: constant(uint256[3]) = [1,2,3]

@external
def foo() -> uint8:
    return MY_CONSTANT[0]
```
```
MY_LIST: constant(uint256[6]) = [1,2,3]

@external
def ix(i: uint256) -> uint8:
    return MY_LIST[i]
```

The bug does not appear to be limited to constant arrays, however. The following code snippet also compiles when it should not:
```
MY_CONSTANT: constant(uint256) = 1

@external
def foo() -> uint8:
    return MY_CONSTANT
```

### How I did it

Propagate the type definition of a constant when it is folded (i.e. replaced by a new AST node) by appending this information as a string to the new node's `_metadata` attribute using the `type` key. This value is then used to narrow down the possible type of a node from a list of possible types. 

### How to verify it

See tests.

### Description for the changelog

Fix type checking for constants

### Cute Animal Picture

Put a link to a cute animal picture inside the parenthesis-->

### CODE REVIEW

File: vyper/ast/folding.py [hunk 1]
```diff
@@ -172,20 +172,39 @@ def replace_user_defined_constants(vyper_module: vy_ast.Module) -> int:
             # annotation is not wrapped in `constant(...)`
             continue
 
-        changed_nodes += replace_constant(vyper_module, node.target.id, node.value, False)
+        # Extract typedef for propagation to type checker
+        node_annotation_arg = node.get("annotation.args")[0]
+        type = None
```
Comment (charles-cooper) [inline vyper/ast/folding.py:177]: `type` is a keyword in python, please use `type_` or something similar

File: vyper/ast/folding.py [hunk 2]
```diff
@@ -172,20 +172,39 @@ def replace_user_defined_constants(vyper_module: vy_ast.Module) -> int:
...
+        )
 
     return changed_nodes
 
 
 # TODO constant folding on log events
 
 
-def _replace(old_node, new_node):
+def _replace(old_node, new_node, type=None):
```
Comment (charles-cooper) [inline vyper/ast/folding.py:196]: same here, please use `type_`

File: vyper/ast/folding.py [hunk 3]
```diff
@@ -172,20 +172,39 @@ def replace_user_defined_constants(vyper_module: vy_ast.Module) -> int:
...
     if isinstance(new_node, vy_ast.Constant):
-        return new_node.from_node(old_node, value=new_node.value)
+        new_node = new_node.from_node(old_node, value=new_node.value)
+        if type:
+            new_node._metadata["type"] = type
```
Comment (charles-cooper) [inline vyper/ast/folding.py:200]: isn't `node._metadata["type"]` usually some kind of type object, not a string?

File: vyper/ast/folding.py [hunk 4]
```diff
@@ -172,20 +172,39 @@ def replace_user_defined_constants(vyper_module: vy_ast.Module) -> int:
             # annotation is not wrapped in `constant(...)`
             continue
 
-        changed_nodes += replace_constant(vyper_module, node.target.id, node.value, False)
+        # Extract typedef for propagation to type checker
+        node_annotation_arg = node.get("annotation.args")[0]
+        type = None
+        if node_annotation_arg.ast_type == "Subscript":
```
Comment (charles-cooper) [inline vyper/ast/folding.py:178]: I don't think you have to parse the type manually here, we have machinery for that

**Review by charles-cooper — CHANGES_REQUESTED** [review]
@tserg thanks for the contribution! I think this PR will be cleaner if you reuse existing machinery for parsing types, and then follow the convention of putting type objects (e.g. some kind of `TypeDefinition`) into `node._metadata["type"]`

File: vyper/ast/folding.py (hunk 3 above)
Comment (tserg) [reply vyper/ast/folding.py:200]: Yes, it is usually a `TypeDefinition`. My initial thinking was that `TypeDefinition` objects are assigned at a later stage under `semantics`, and it could be inappropriate/messy/premature to import those classes and helper functions into the `ast` module. However, as you pointed out, the inconsistency/messiness ends up in the `_metadata["type"]` instead.

Comment (charles-cooper) [reply vyper/ast/folding.py:200]: Right that's very interesting, I forgot type checking happens after constant folding. We should consider switching the order of the phases.

Comment (tserg) [reply vyper/ast/folding.py:200]: I tried moving the type check phase before constant folding. The good news is that it resolves the issue being addressed in this PR. The bad news is that there are about another 100 test cases that fail. 

Some of the type-checking issues raised are due to:
- Values for built-in constants such as `ZERO_ADDRESS` not being injected, and therefore not being 
- Values for user-defined constants in cases where they are used as subscript or loop ranges not being injected

The remaining issues will require more investigation (e.g. type mismatches, method id must be folded, missing "type" key in metadata).

**Comment by tserg** [comment]
Update: I tried to refactor the parametrizing of distinct integer types for the negative test cases to reduce duplication and potentially in view of a much bigger list of integer types. However, it seems that `pytest` has an existing limitation regarding how fixtures can be used in parametrize. IMO, the best approach at this moment is to rely on the workaround in the first post of this thread (https://github.com/pytest-dev/pytest/issues/349).

File: tests/parser/types/test_lists.py [hunk 5]
```diff
@@ -428,6 +428,45 @@ def ix(i: uint256) -> {type}:
...
+@external
+def foo() -> {return_type}[3]:
+    return MY_CONSTANT
+    """
+    assert_compile_failed(lambda: get_contract(code), InvalidType)
+
+
+def test_constant_list_fail_2(get_contract, assert_compile_failed, distinct_int_types):
+    type = distinct_int_types[0]
+    return_type = distinct_int_types[1]
```
Comment (fubuloubu) [inline tests/parser/types/test_lists.py:446]: Side note: Python supports ["tuple destructuring"](https://blog.teclado.com/destructuring-in-python/)
```suggestion
    type, return_type = distinct_int_types
```

File: tests/conftest.py [hunk 6]
```diff
@@ -208,3 +209,14 @@ def _f(_addr, _salt, _initcode):
...
+
+def get_distinct_integer_types_combinations():
+    # TODO Replace with master list of integer types
+    types_list = ["uint8", "int128", "int256", "uint256"]
+    return [(t, rt) for (t, rt) in product(types_list, types_list) if t != rt]
+
+
+@pytest.fixture(params=get_distinct_integer_types_combinations())
+def distinct_int_types(request):
+    return request.param
```
Comment (fubuloubu) [inline tests/conftest.py:222]: Hmm, yeah this isn't all that great looking, is it?

File: tests/parser/types/numbers/test_constants.py [hunk 7]
```diff
@@ -127,6 +128,19 @@ def test_add(a: uint256) -> uint256:
     assert c.test_add(7) == 40
 
 
+def test_custom_constants_fail(get_contract, assert_compile_failed, distinct_int_types):
```
Comment (fubuloubu) [inline tests/parser/types/numbers/test_constants.py:131]: Yeah, this does feel much harder to read

Let's go back to this if we could:
```py
# Would be nice to put this somewhere accessible, like in vyper.types or something
types = ['uint256', 'uint8', 'int128', 'int256']

@pytest.mark.parametrize("storage_type,return_type", itertools.permutations(types, 2))
```

File: tests/conftest.py (hunk 6 above)
Comment (tserg) [reply tests/conftest.py:222]: Yea, it does make reading the tests more convoluted and having to jump through multiple files.

File: vyper/ast/folding.py [hunk 8]
```diff
@@ -172,20 +174,40 @@ def replace_user_defined_constants(vyper_module: vy_ast.Module) -> int:
             # annotation is not wrapped in `constant(...)`
             continue
 
-        changed_nodes += replace_constant(vyper_module, node.target.id, node.value, False)
+        # Extract typedef for propagation to type checker
+        node_annotation_arg = node.get("annotation.args")[0]
+        type_: Union[BaseTypeDefinition, None] = None
+        try:
+            type_ = get_type_from_annotation(node_annotation_arg, DataLocation.STORAGE)
+        except:  # noqa: E722
```
Comment (fubuloubu) [inline vyper/ast/folding.py:182]: Can we except `VyperException`?
https://github.com/vyperlang/vyper/blob/master/vyper/exceptions.py#L29

**Review by fubuloubu — APPROVED** [review]

**Review by fubuloubu — APPROVED** [review]

File: vyper/ast/folding.py (hunk 3 above)
Comment (charles-cooper) [reply vyper/ast/folding.py]: In that case, we should not use the `"type"` key because having `"type"` be polymorphic is really messy. We should think about a more long-term solution but for now I think we can attach a `"constant_annotation"`, which we can inspect during the annotation / type checking phase. Also I would just attach the entire annotation (AST node) instead of converting it to a string.

**Review by fubuloubu — APPROVED** [review]

File: vyper/semantics/validation/utils.py [hunk 9]
```diff
@@ -115,8 +117,21 @@ def get_possible_types_from_node(self, node, only_definitions=True):
...
 
         # Restrict possible types if typedef is propagated in metadata
-        if "type" in node._metadata:
-            propagated_type = node._metadata["type"]
+        if "constant_annotation" in node._metadata:
+            # TODO investigate this circular dependency
+            from vyper.semantics.types.utils import get_type_from_annotation
+
+            propagated_type = get_type_from_annotation(
+                node._metadata["constant_annotation"], DataLocation.STORAGE
```
Comment (charles-cooper) [inline vyper/semantics/validation/utils.py:125]: maybe `DataLocation.UNSET`

File: vyper/semantics/validation/utils.py [hunk 10]
```diff
@@ -113,6 +115,28 @@ def get_possible_types_from_node(self, node, only_definitions=True):
...
+            )
+
+            if not isinstance(node, vy_ast.List) and isinstance(
+                propagated_type, (ArrayDefinition, DynamicArrayDefinition)
+            ):
+
+                # For array elements, extract the base type from the array definition
+                propagated_type = propagated_type.value_type
+
+            for _, t in enumerate(types_list):
```
Comment (charles-cooper) [inline vyper/semantics/validation/utils.py:135]: `enumerate` not needed here

File: vyper/semantics/validation/utils.py [hunk 11]
```diff
@@ -113,6 +115,28 @@ def get_possible_types_from_node(self, node, only_definitions=True):
...
+                if type(propagated_type) is type(t):
```
Comment (charles-cooper) [inline vyper/semantics/validation/utils.py:136]: direct type comparison always jumps out at me .. can you help me understand what the intention of this loop is?

Comment (tserg) [reply vyper/semantics/validation/utils.py:136]: The intention is to ensure that the propagated type is one of the possible types retrieved for that node so as to maintain some logical consistency with what the `get_possible_types_from_node` function is already doing, and also potentially as a failsafe in some edge case where the propagated type may not be one of the possible types retrieved.

The test suite still passes if we replace this loop with the following statement:
```
types_list = [propagated_type]
```
If this is the preferred option, then we can shift this segment for constant typedef propagation to the beginning of the function so that the propagated type is taken to be the default possible type and returned immediately. There would be no need to retrieve the possible types for this node. We can view this function as then differentiating between nodes with constant typedef propagation metadata and nodes without.

Comment (tserg) [reply vyper/semantics/validation/utils.py:136]: An alternative to type comparison in the existing approach is to compare the `__repr__()` string representations.

**Review by charles-cooper — APPROVED** [review]
